src/proof/er/er_proof.cpp

```diff
@@ -0,0 +1,378 @@
+ **/
+
+#include "proof/er/er_proof.h"
+
+#include <unistd.h>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <unordered_set>
+
+#include "base/cvc4_assert.h"
+#include "base/map_util.h"
+#include "proof/dimacs_printer.h"
+#include "proof/lfsc_proof_printer.h"
+#include "proof/proof_manager.h"
+
+#if CVC4_USE_DRAT2ER
+#include "drat2er.h"
+#include "drat2er_options.h"
+#endif
+
+namespace CVC4 {
+namespace proof {
+namespace er {
+
+TraceCheckProof TraceCheckProof::fromText(std::istream& in)
+{
+  TraceCheckProof pf;
+  TraceCheckIdx idx = 0;
+  int64_t token = 0;
+
+  // For each line of the proof, start with the idx
+  // If there is no idx, then you're done!
+  in >> idx;
+  for (; !in.eof(); in >> idx)
+  {
+    Assert(in.good());
+
+    // Then parse the clause (it's 0-terminated)
+    std::vector<prop::SatLiteral> clause;
+    in >> token;
+    for (; token != 0; in >> token)
+    {
+      clause.emplace_back(std::abs(token) - 1, token < 0);
+    }
+
+    // Then parse the chain of literals (it's also 0-terminated)
+    std::vector<TraceCheckIdx> chain;
+    in >> token;
+    for (; token != 0; in >> token)
+    {
+      Assert(token > 0);
+      chain.push_back(token);
+    }
+
+    // Add the line to the proof
+    pf.d_lines.emplace_back(idx, std::move(clause), std::move(chain));
+  }
+  return pf;
+}
+
+ErProof ErProof::fromBinaryDratProof(const ClauseUseRecord& usedClauses,
+                                     const std::string& dratBinary)
+{
+  std::ostringstream cmd;
+  char formulaFilename[] = "/tmp/cvc4-dimacs-XXXXXX";
+  char dratFilename[] = "/tmp/cvc4-drat-XXXXXX";
+  char tracecheckFilename[] = "/tmp/cvc4-tracecheck-er-XXXXXX";
+
+  int r;
+  r = mkstemp(formulaFilename);
+  AlwaysAssert(r > 0);
+  close(r);
+  r = mkstemp(dratFilename);
+  AlwaysAssert(r > 0);
+  close(r);
+  r = mkstemp(tracecheckFilename);
+  AlwaysAssert(r > 0);
+  close(r);
+
+  // Write the formula
+  std::ofstream formStream(formulaFilename);
+  printDimacs(formStream, usedClauses);
+  formStream.close();
+
+  // Write the (binary) DRAT proof
+  std::ofstream dratStream(dratFilename);
+  dratStream << dratBinary;
+  dratStream.close();
+
+  // Invoke drat2er
+#if CVC4_USE_DRAT2ER
+  drat2er::TransformDRATToExtendedResolution(formulaFilename,
+                                             dratFilename,
+                                             tracecheckFilename,
+                                             false,
+                                             drat2er::options::QUIET);
+#else
+  Unimplemented(
+      "ER proof production requires drat2er.\n"
+      "Run contrib/get-drat2er, reconfigure with --drat2er, and rebuild");
+#endif
+
+  // Parse the resulting TRACECHECK proof into an ER proof.
+  std::ifstream tracecheckStream(tracecheckFilename);
+  ErProof proof(usedClauses, TraceCheckProof::fromText(tracecheckStream));
+  tracecheckStream.close();
+
+  unlink(tracecheckFilename);
```

Comment:
I would suggest to unlink the files as soon as they are opened. Like that, they are deleted as soon as CVC4 terminates, no matter the reason (see #2834).

Reply:
I see what you mean. Good idea, will do.

Reply:
To ensure that the file isn't removed while the program is running, I delayed our file closes.

Reply:
The file itself is only removed after the last program that has it open is terminated [[0]](http://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html). However, I just realized there is an issue with it because while it is guaranteed that the file is not deleted, the file is not accessible through the unlinked path anymore. Let's leave it as it is for now.